File: sdks/python/src/novelcraft/sdk/section.py
from typing import List

from .position import Position
from .block import Block


class Section:
    def __init__(self, position: Position[int], block_id_list: List[int]):
        if position.x % 16 != 0 or position.y % 16 != 0 or position.z % 16 != 0:
            raise ValueError("The position must be a multiple of 16.")

        if len(block_id_list) != 4096:
            raise ValueError("The block ID list must have a length of 4096.")

        self._block_id_list = block_id_list
        self._position = position

    def __getitem__(self, position: Position[int]) -> Block:
        if position.x < 0 or position.x > 15 \
           or position.y < 0 or position.y > 15 \
           or position.z < 0 or position.z > 15:
            raise IndexError("The position is out of range.")

        return Block(self._block_id_list[position.x * 256 + position.y * 16 + position.z],
                      position)

    def __setitem__(self, position: Position[int], block: Block):
        if position.x < 0 or position.x > 15 \
           or position.y < 0 or position.y > 15 \
           or position.z < 0 or position.z > 15:
            raise IndexError("The position is out of range.")

        self._block_id_list[position.x * 256 + position.y * 16 + position.z] = block.get_type_id()

    def get_position(self) -> Position[int]:
        return self._position

File: sdks/dotnet/src/Section.cs
namespace NovelCraft.Sdk;

internal class Section {
  public IBlock this[IPosition<int> relativePosition] {
    get {
      // Check if the position is valid.
      if (relativePosition.X < 0 || relativePosition.X > 15 ||
            relativePosition.Y < 0 || relativePosition.Y > 15 ||
            relativePosition.Z < 0 || relativePosition.Z > 15) {
        throw new ArgumentException("The position should be in range [0, 15].");
      }

      return new Block(
        _blockIdList[relativePosition.X * 256 + relativePosition.Y * 16 + relativePosition.Z],
        new Position<int>(Position.X + relativePosition.X, Position.Y + relativePosition.Y, Position.Z + relativePosition.Z)
        );
    }

    set {
      // Check if the position is valid.
      if (relativePosition.X < 0 || relativePosition.X > 15 ||
            relativePosition.Y < 0 || relativePosition.Y > 15 ||
            relativePosition.Z < 0 || relativePosition.Z > 15) {
        throw new ArgumentException("The position should be in range [0, 15].");
      }

      _blockIdList[relativePosition.X * 256 + relativePosition.Y * 16 + relativePosition.Z] = value.TypeId;
    }
  }

  public IPosition<int> Position { get; }

  private List<int> _blockIdList;


  public Section(IPosition<int> position, List<int> blockIdList) {
    // Check if the position is valid.
    if (position.X % 16 != 0 || position.Y % 16 != 0 || position.Z % 16 != 0) {
      throw new ArgumentException("The position should be multiples of 16.");
    }

    // Check if the block ID list is right size.
    if (blockIdList.Count != 4096) {
      throw new ArgumentException("The block ID list must have 4096 elements.");
    }

    Position = new Position<int>(position);
    _blockIdList = new(blockIdList);
  }
}